Use cached client for apply

tim-ebert opened the pull request:
**What this PR does / why we need it**:
This PR adds some more performance (and networking) optimisations, which become relevant, when using grm for quite large ManagedResources (like the `seed-bootstrap` chart).
1. It now also uses the caching target client for get calls before applying changes to the objects
2. It reduces the amount of unnecessary update calls by retaining some more metadata while merging and by using semantic equality to check if objects are changed.
3. It also fixes a problem, which caused mr reconciliation to break, if at least one of the objects cannot be mapped by the target rest mapper.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy|action
- target_group:   user|operator|developer
-->
```improvement operator
NONE
```

rfranzke commented on pkg/controller/managedresources/controller.go:
```diff
@@ -187,23 +187,25 @@ func (r *Reconciler) reconcile(mr *resourcesv1alpha1.ManagedResource, log logr.L
-						r.targetRESTMapper.Reset()
-					}
-					log.Error(err, fmt.Sprintf("could not get rest mapping for %s '%s/%s'", obj.GetKind(), obj.GetNamespace(), obj.GetName()),
+					// Don't reset RESTMapper in case of cache misses. Most probably indicates, that the corresponding CRD is not yet applied.
```
Can you explain why the reset is not needed anymore? I see it's also done below, but I don't have the context about the relation anymore.

tim-ebert replied:
Sure :)
At this point, the rest mapper is used to figure out the scope of the to-be-applied-object. Before, we were defaulting the object's namespace to `default` if the resource is not a `Namespace`, which then was leading to many references in the object status like this:
```yaml
  - apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    name: foo
    namespace: default
```

The healthcheck controller later picks up those references to get the objects from the api server. Before using the cached client for that, it wasn't a problem to specify a namespace in a key for a non-namespaced object. But the cached client can't handle that, so I figured, we should properly check the objects scope before defaulting the namespace field.

The problem now is, that if you create a ManagedResource containing a CRD (e.g. VPA) and a lot of corresponding CRs, the target mapper is reset for each CR, if the CRD has not been created yet (e.g. `seed-bootstrap` applied for the first time). This will therefore lead to a lot of discovery requests, which could even be rate limited in extreme cases. That's why I am not resetting the rest mapper here anymore.
Requests to create such CRs will anyways fail at the first try, until the CRD has been accepted, then the rest mapper will be reset and in the next try, it will probably succeed.
But as I am writing this, I figured, I should probably reset the mapper only once per reconcile down below instead of for each cache miss for the exact same reason. 😄

rfranzke replied:
Alright, thank you for the detailed explanation! 
> But as I am writing this, I figured, I should probably reset the mapper only once per reconcile down below instead of for each cache miss for the exact same reason. 😄

Sounds good :)

rfranzke commented on pkg/controller/managedresources/merger.go:
```diff
@@ -39,6 +39,12 @@ func merge(desired, current *unstructured.Unstructured, forceOverwriteLabels boo
+	newObject.SetCreationTimestamp(oldObject.GetCreationTimestamp())
+	newObject.SetSelfLink(oldObject.GetSelfLink())
+	newObject.SetUID(oldObject.GetUID())
+	newObject.SetGeneration(oldObject.GetGeneration())
```
There are also other fields in the `metadata`, e.g. `clusterName`, `managedFields`, etc. What about them? Wouldn't it be safer to copy the whole metadata and then simply overwrite the fields that we want to overwrite? Though, not sure if that's possible?

tim-ebert replied:
Valid point, I changed it to copying the whole `metadata` stanza and tested it/played around with it a wile and it seems to work fine :)

rfranzke commented on pkg/controller/utils/utils.go:
```diff
@@ -138,6 +140,67 @@ func TryUpdate(ctx context.Context, backoff wait.Backoff, c client.Client, obj r
+// API server, applies the given mutate func and creates or updates it afterwards. In contrast to
+// controllerutil.CreateOrUpdate it tries to create a new typed object of obj's kind (using the provided scheme)
+// to make typed Get requests in order to leverage the client's cache.
+func TypedCreateOrUpdate(ctx context.Context, scheme *runtime.Scheme, c client.Client, obj *unstructured.Unstructured, mutate func() error) error {
```
nice :) Does a unit test make sense for this function?

tim-ebert replied:
Yes, done 👍

tim-ebert commented:
I have updated the PR, PTAL :)

Updates:
- Refactored `mergeMapsBasedOnOldMap()` a bit to make it easier to understand and added some unit tests for merging labels and annotations. I hope, I didn't change the logic in any special case. @rfranzke can you please check?
- Added a unit test for `TypedCreateOrUpdate`
- Added logic and unit tests to copy all metadata in `merger.go`
- Only reset the `targetRESTMapper` once per reconcile to avoid multiple discovery calls.

rfranzke commented on pkg/controller/managedresources/merger.go:
```diff
@@ -36,8 +36,15 @@ func merge(desired, current *unstructured.Unstructured, forceOverwriteLabels boo
+		// cast to map to be able to check if metadata is empty
+		if oldMetadataMap, ok := oldMetadataInterface.(map[string]interface{}); ok {
+			if len(oldMetadataMap) > 0 {
+				newObject.Object["metadata"] = oldMetadataMap
```
It seems like this was also the case before, but would this approach work if we change something in the metadata, e.g., we add a finalizer to the manifests? It would get overwritten here, or? Do we need to support such cases?

tim-ebert replied:
Indeed, if you would change metadata other than `labels` or `annotations` it would get overwritten.
I double checked in one of our soils and we don't have any metadata other than `name,namespace,labels,annotations` in all secrets used by managed resources.

As you said, adding a finalizer to the manifest was not supported before and I can't think of a case, where this is actually needed. So I guess, this should be fine, but if we encounter a case where this is a valid requirement, it should be easy to add here :)

rfranzke replied:
Yes, sure, alright, thanks for clarifying!
